File: samples/vdsp_op/resize/resize.py
import vaststreamx as vsx
import numpy as np
import cv2
import argparse
import ast

attr = vsx.AttrKey


def argument_parser():
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "-d",
        "--device_id",
        default=0,
        type=int,
        help="device id to run",
    )
    parser.add_argument(
        "--input_file",
        default="data/images/dog.jpg",
        help="input image",
    )
    parser.add_argument(
        "--output_size", help="resize output size", default="[256, 256]"
    )
    parser.add_argument(
        "--output_file",
        default="./resize_result.jpg",
        help="output image",
    )
    args = parser.parse_args()
    return args


def resize_bgr888_to_bgr888(args):
    """
    resize image, output image format is the same as input image format
    """
    cv_image = cv2.imread(args.input_file)
    assert cv_image is not None, f"Failed to read {args.input_file}"
    vsx_image = vsx.create_image(
        cv_image,
        vsx.ImageFormat.BGR_INTERLEAVE,
        cv_image.shape[1],
        cv_image.shape[0],
        args.device_id,
    )
    # Build operator
    op = vsx.BuildInOperator(vsx.OpType.SINGLE_OP_RESIZE)

    output_size = ast.literal_eval(args.output_size)
    # set attri
    out_width = output_size[0]
    out_height = output_size[1]

    op.set_attribute(
        {
            attr.IIMAGE_WIDTH: vsx_image.width,
            attr.IIMAGE_HEIGHT: vsx_image.height,
            attr.IIMAGE_WIDTH_PITCH: vsx_image.width,
            attr.IIMAGE_HEIGHT_PITCH: vsx_image.height,
            attr.OIMAGE_WIDTH: out_width,
            attr.OIMAGE_HEIGHT: out_height,
            attr.OIMAGE_WIDTH_PITCH: out_width,
            attr.OIMAGE_HEIGHT_PITCH: out_height,
            attr.IIMAGE_FORMAT: vsx.ImageType.RGB888,
            attr.OIMAGE_FORMAT: vsx.ImageType.RGB888,
            attr.RESIZE_TYPE: vsx.ImageResizeType.BILINEAR_CV,
        }
    )

    # run
    output_list = op.execute(
        [vsx_image], [(vsx.ImageFormat.BGR_INTERLEAVE, out_width, out_height)]
    )

    out_image_np = vsx.as_numpy(output_list[0]).squeeze()
    # write
    out_image = out_image_np.reshape((out_height, out_width, 3))
    cv2.imwrite(args.output_file, out_image)
    print("save result to ", args.output_file)


def resize_bgr888_to_bgr_planar(args):
    """
    resize image, and cvtcolor from bgr888 to bgr_planar
    """
    cv_image = cv2.imread(args.input_file)
    assert cv_image is not None, f"Failed to read {args.input_file}"
    vsx_image = vsx.create_image(
        cv_image,
        vsx.ImageFormat.BGR_INTERLEAVE,
        cv_image.shape[1],
        cv_image.shape[0],
        args.device_id,
    )
    # Build operator
    op = vsx.BuildInOperator(vsx.OpType.SINGLE_OP_RESIZE)

    output_size = ast.literal_eval(args.output_size)
    # set attri
    out_width = output_size[0]
    out_height = output_size[1]

    op.set_attribute(
        {
            attr.IIMAGE_WIDTH: vsx_image.width,
            attr.IIMAGE_HEIGHT: vsx_image.height,
            attr.IIMAGE_WIDTH_PITCH: vsx_image.width,
            attr.IIMAGE_HEIGHT_PITCH: vsx_image.height,
            attr.OIMAGE_WIDTH: out_width,
            attr.OIMAGE_HEIGHT: out_height,
            attr.OIMAGE_WIDTH_PITCH: out_width,
            attr.OIMAGE_HEIGHT_PITCH: out_height,
            attr.IIMAGE_FORMAT: vsx.ImageType.RGB888,
            attr.OIMAGE_FORMAT: vsx.ImageType.RGB_PLANAR,
            attr.RESIZE_TYPE: vsx.ImageResizeType.BILINEAR_CV,
        }
    )

    # run
    output_list = op.execute(
        [vsx_image], [(vsx.ImageFormat.BGR_PLANAR, out_width, out_height)]
    )

    out_image_np = vsx.as_numpy(output_list[0]).squeeze()
    # write
    out_image = out_image_np.transpose((1, 2, 0))  # BGR_PLANAR to BGR_INTERLEAVE
    cv2.imwrite(args.output_file, out_image)
    print("save result to ", args.output_file)


if __name__ == "__main__":
    args = argument_parser()
    vsx.set_device(args.device_id)
    resize_bgr888_to_bgr888(args)
    # resize_bgr888_to_bgr_planar(args)

File: samples/vdsp_op/resize/resize.cpp

/*
 * Copyright (C) 2025 Vastai-tech Company.
 * All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
#include "common/cmdline.hpp"
#include "common/utils.hpp"
#include "opencv2/opencv.hpp"
#include "vaststreamx/vaststreamx.h"

cmdline::parser ArgumentParser(int argc, char** argv) {
  cmdline::parser args;
  args.add<uint32_t>("device_id", 'd', "device id to run", false, 0);
  args.add<std::string>("input_file", '\0', "input image", false,
                        "../data/images/dog.jpg");
  args.add<std::string>("output_size", '\0', "output size", false, "[256,256]");
  args.add<std::string>("output_file", '\0', "output image", false,
                        "resize_result.jpg");
  args.parse_check(argc, argv);
  return args;
}
int resize_bgr888_to_bgr888(cmdline::parser& args) {
  uint32_t device_id = args.get<uint32_t>("device_id");
  vsx::Image image;
  vsx::MakeVsxImage(args.get<std::string>("input_file"), image,
                    vsx::RGB_INTERLEAVE);

  std::vector<uint32_t> output_size =
      vsx::ParseVecUint(args.get<std::string>("output_size"));
  int output_width = output_size[0];
  int output_height = output_size[1];

  auto op = vsx::BuildInOperator(vsx::BuildInOperatorType::kSINGLE_OP_RESIZE);

  op.SetAttribute<vsx::AttrKey::kIimageWidth>(image.Width());
  op.SetAttribute<vsx::AttrKey::kIimageHeight>(image.Height());
  op.SetAttribute<vsx::AttrKey::kIimageWidthPitch>(image.Width());
  op.SetAttribute<vsx::AttrKey::kIimageHeightPitch>(image.Height());
  op.SetAttribute<vsx::AttrKey::kIimageFormat>(
      vsx::BuildInOperatorAttrImageType::kRGB888);
  op.SetAttribute<vsx::AttrKey::kOimageWidth>(output_width);
  op.SetAttribute<vsx::AttrKey::kOimageHeight>(output_height);
  op.SetAttribute<vsx::AttrKey::kOimageWidthPitch>(output_width);
  op.SetAttribute<vsx::AttrKey::kOimageHeightPitch>(output_height);
  op.SetAttribute<vsx::AttrKey::kOimageFormat>(
      vsx::BuildInOperatorAttrImageType::kRGB888);
  op.SetAttribute<vsx::AttrKey::kResizeType>(
      vsx::ImageResizeType::kRESIZE_TYPE_BILINEAR_CV);

  // copy input to device memory
  auto input_vacc = vsx::Image(image.Format(), image.Width(), image.Height(),
                               vsx::Context::VACC(device_id));
  input_vacc.CopyFrom(image);

  // make device memory image for output
  auto output_vacc = vsx::Image(image.Format(), output_width, output_height,
                                vsx::Context::VACC(device_id));
  std::vector<vsx::Image> outputs{output_vacc};

  // buildin operator execute
  op.Execute({input_vacc}, outputs);

  // copy result to cpu memory
  vsx::Image output_cpu = vsx::Image(output_vacc.Format(), output_vacc.Width(),
                                     output_vacc.Height());
  output_cpu.CopyFrom(output_vacc);

  cv::Mat out_mat;
  vsx::ConvertVsxImageToCvMatBgrPacked(output_cpu, out_mat);
  cv::imwrite(args.get<std::string>("output_file"), out_mat);
  std::cout << "Save result to " << args.get<std::string>("output_file")
            << std::endl;

  return 0;
}

int resize_rgb888_to_rgb_planar(cmdline::parser& args) {
  uint32_t device_id = args.get<uint32_t>("device_id");
  vsx::Image image;
  vsx::MakeVsxImage(args.get<std::string>("input_file"), image,
                    vsx::RGB_INTERLEAVE);

  std::vector<uint32_t> output_size =
      vsx::ParseVecUint(args.get<std::string>("output_size"));
  int output_width = output_size[0];
  int output_height = output_size[1];

  auto op = vsx::BuildInOperator(vsx::BuildInOperatorType::kSINGLE_OP_RESIZE);

  op.SetAttribute<vsx::AttrKey::kIimageWidth>(image.Width());
  op.SetAttribute<vsx::AttrKey::kIimageHeight>(image.Height());
  op.SetAttribute<vsx::AttrKey::kIimageWidthPitch>(image.Width());
  op.SetAttribute<vsx::AttrKey::kIimageHeightPitch>(image.Height());
  op.SetAttribute<vsx::AttrKey::kIimageFormat>(
      vsx::BuildInOperatorAttrImageType::kRGB888);
  op.SetAttribute<vsx::AttrKey::kOimageWidth>(output_width);
  op.SetAttribute<vsx::AttrKey::kOimageHeight>(output_height);
  op.SetAttribute<vsx::AttrKey::kOimageWidthPitch>(output_width);
  op.SetAttribute<vsx::AttrKey::kOimageHeightPitch>(output_height);
  op.SetAttribute<vsx::AttrKey::kOimageFormat>(
      vsx::BuildInOperatorAttrImageType::kRGB_PLANAR);
  op.SetAttribute<vsx::AttrKey::kResizeType>(
      vsx::ImageResizeType::kRESIZE_TYPE_BILINEAR_CV);

  // copy input to device memory
  auto input_vacc = vsx::Image(image.Format(), image.Width(), image.Height(),
                               vsx::Context::VACC(device_id));
  input_vacc.CopyFrom(image);

  // make device memory image for output
  auto output_vacc = vsx::Image(vsx::RGB_PLANAR, output_width, output_height,
                                vsx::Context::VACC(device_id));
  std::vector<vsx::Image> outputs{output_vacc};

  // buildin operator execute
  op.Execute({input_vacc}, outputs);

  // copy result to cpu memory
  vsx::Image output_cpu = vsx::Image(output_vacc.Format(), output_vacc.Width(),
                                     output_vacc.Height());
  output_cpu.CopyFrom(output_vacc);

  cv::Mat out_mat;
  vsx::ConvertVsxImageToCvMatBgrPacked(output_cpu, out_mat);
  cv::imwrite(args.get<std::string>("output_file"), out_mat);
  std::cout << "Save result to " << args.get<std::string>("output_file")
            << std::endl;
  return 0;
}

int main(int argc, char* argv[]) {
  auto args = ArgumentParser(argc, argv);
  vsx::SetDevice(args.get<uint32_t>("device_id"));

  resize_bgr888_to_bgr888(args);
  // resize_rgb888_to_rgb_planar(args);
  return 0;
}